Recommend using verbatim string literals for regexes

yawaramin opened the pull request:
To avoid the need to escape backslashes.

yawaramin committed:
Recommend using verbatim string literals for regexes

To avoid the need to escape backslashes.

gasche commented on otherlibs/str/str.mli:
```diff
@@ -50,21 +50,20 @@ val regexp : string -> regexp
+   easier to use a verbatim string literal [{|...|}] to avoid having to
+   escape backslashes. And, if you want a regular expression that
+   matches a literal backslash character, you need to double it, i.e.
+   [Str.regexp {|\\|}].
```
The "And [...]" sentence is unrelatd to the example below, so it should come after the example (as in the original version). Could you also avoid `i.e.` which is jargon?

gasche commented on otherlibs/str/str.mli:
```diff
@@ -50,21 +50,20 @@ val regexp : string -> regexp
+   For example, the following expression:
+   {[ let r = Str.regexp {|hello \([A-Za-z]+\)|} in
+      Str.replace_first r {|\1|} "hello world" ]}
    returns the string ["world"].
```
Could you show the example twice, once with usual string literals and the other with quoted string literals?

gasche commented on otherlibs/str/str.mli:
```diff
@@ -50,21 +50,20 @@ val regexp : string -> regexp
-*)
+   This is just a recommendation; you can use double quote string
+   literals i.e.  ["..."], however you will have to carefully escape
+   backslashes. *)
```
This would not be necessary anymore if the example was shown in two versions.

gasche commented on otherlibs/str/str.mli:
```diff
@@ -50,21 +50,20 @@ val regexp : string -> regexp
-   {[ let r = Str.regexp "hello \\([A-Za-z]+\\)" in
-      Str.replace_first r "\\1" "hello world" ]}
+   In regular expressions you will often use backslash characters; it's
+   easier to use a verbatim string literal [{|...|}] to avoid having to
```
The OCaml manual calls them "quoted string literals", not "verbatim string literals". ("quoted" is maybe confusing here, but let's use the standard name anyway.)

yawaramin committed:
Address review comments

- Use standard naming 'quoted string literals'
- Show the same example using both types of string literals

yawaramin commented:
@gasche thank you for the review. I believe I have addressed all the comments.

Octachron commented on otherlibs/str/str.mli:
```diff
@@ -50,21 +50,25 @@ val regexp : string -> regexp
+
+   You can use regular string literals ["..."] too, however you will
+   have to escape backslashes. To repeat the example above with a
+   regular string literal:
```
There is a missing verb. Maybe

> The example above can be rewritten with a regular string literal as

?

Octachron commented on otherlibs/str/str.mli:
```diff
@@ -50,21 +50,25 @@ val regexp : string -> regexp
-   syntax of string constants in OCaml) by doubling them again, so you
-   need to write four backslash characters: [Str.regexp "\\\\"].
-*)
+   And the double backslash regular expression: [Str.regexp "\\\\"]. *)
```
Similarly, with a verb and repeating the objective of the example

> And the regular expression for matching a backslash becomes a quadruple backslash:  [Str.regexp "\\\\"].

gasche reviewed (approved):
I like the new version, thanks!

(Personally I'm fine with the sentences that don't have a verb, but whatever, let's also make @Octachron happy.)

Octachron commented:
Note that if my remarks are too bothersome, I am utterly fine with merging in the current state.

gasche commented:
Let's wait a bit more to see if @yawaramin has feedback, otherwise we'll just merge.

yawaramin committed:
Include octachron's suggestions

yawaramin commented:
@Octachron thanks for the suggestions–I agree your wording is better. Updated!

gasche commented:
Thanks! Merged, and included in 4.14 as 4a3f29d57a3234c473a121428c0cdcf37b7df82d.
